The most important aspect before **any application** of Machine Learning is undoubtedly the pre-processing of data.

Several books, blog posts, tutorials are concern with the nice part of the job (e.g. insights of the data, applying novel algorithms, et cetera) or the application for it self of the algorithms(e.g. Linear Regression, Support Vector Machines, Neural Networks, et cetera ...).

Nevertheless the dataset (A.K.A. dataset, database, et cetera) plays an important decisive role in the success or lack thereof in the implementation of Machine Learning.

In a quick analogy, the data is like a key and algorithms are like a lock of a door.

You may have a key with some friction  ... (e.g. truncation data, missing data, non-standard data, et cetera)

You may have the door lock with some friction ... (e.g. parametric adjustment for 'fitting', eliminating steps of calculations, using 'ensembles'algorithms in the pipeline, et cetera)

But the lock will only open with the corresponding key. (e.g. database with the attributes of prediction and outcome (target) correct).

Now that we know the importance of having the data to application of algorithms, and consequently creating models, let's talk about an important aspect of the issue of input data that is the Pre-processing.

The pre-processing of data is any activity on the adjustment of the data both in terms of format, sampling so that the algorithms are able to perform construction activities of the models in order to have a better use computationally (temporal complexity + complexity space).

The main activities of the preprocessing of data are:


- Amostragem (Sampling)
- Feature Engineering que contém (i) Feature Extraction e (ii) Feature Selection
- Binning 
- Scaling
- Padronização (Standarize)
- Normalização
- Dimensionality Reduction

First at all, lets import the dataset of scikit package. 

In [31]:
from sklearn import datasets

Now, let's create an object called *datasets* and load Iris dataset.

In [2]:
datasets = datasets.load_iris()

This dataset is splited in two sets, one called data and another called target.
O conjunto de dados 'data' possuí todos os atributos (i.e. variáveis independentes, predictors, features, etc) relativos à base Iris.

The set target have all classes of dataset Iris. 

Let's see the quantity of data in each dataset using the function shape. 

In [33]:
print(datasets.data.shape)

AttributeError: module 'sklearn.datasets' has no attribute 'data'

In resultset data we have 150 records and 4 attributes.

Let's see qhat we have in dataset target.

In [34]:
print(datasets.target.shape)

AttributeError: module 'sklearn.datasets' has no attribute 'target'

As we can see, we have only a column that is the dependent variable. 

Best pratice: Always look to your data. Always.

In [32]:
datasets.data

AttributeError: module 'sklearn.datasets' has no attribute 'data'

In [6]:
datasets.target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

For explamplification we'll put all data in another objects. 

For naming proposes every time that we use the predictor variables we'll use X as object, and y for dependent variable in target dataset. 

In [7]:
X = datasets.data

In [8]:
y = datasets.target

Let's see the data in X and y objects.

In [9]:
X

array([[ 5.1,  3.5,  1.4,  0.2],
       [ 4.9,  3. ,  1.4,  0.2],
       [ 4.7,  3.2,  1.3,  0.2],
       [ 4.6,  3.1,  1.5,  0.2],
       [ 5. ,  3.6,  1.4,  0.2],
       [ 5.4,  3.9,  1.7,  0.4],
       [ 4.6,  3.4,  1.4,  0.3],
       [ 5. ,  3.4,  1.5,  0.2],
       [ 4.4,  2.9,  1.4,  0.2],
       [ 4.9,  3.1,  1.5,  0.1],
       [ 5.4,  3.7,  1.5,  0.2],
       [ 4.8,  3.4,  1.6,  0.2],
       [ 4.8,  3. ,  1.4,  0.1],
       [ 4.3,  3. ,  1.1,  0.1],
       [ 5.8,  4. ,  1.2,  0.2],
       [ 5.7,  4.4,  1.5,  0.4],
       [ 5.4,  3.9,  1.3,  0.4],
       [ 5.1,  3.5,  1.4,  0.3],
       [ 5.7,  3.8,  1.7,  0.3],
       [ 5.1,  3.8,  1.5,  0.3],
       [ 5.4,  3.4,  1.7,  0.2],
       [ 5.1,  3.7,  1.5,  0.4],
       [ 4.6,  3.6,  1. ,  0.2],
       [ 5.1,  3.3,  1.7,  0.5],
       [ 4.8,  3.4,  1.9,  0.2],
       [ 5. ,  3. ,  1.6,  0.2],
       [ 5. ,  3.4,  1.6,  0.4],
       [ 5.2,  3.5,  1.5,  0.2],
       [ 5.2,  3.4,  1.4,  0.2],
       [ 4.7,  3.2,  1.6,  0.2],
       [ 4

In [10]:
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

Para importar dados que estejam na web, podemos usar o numpy que é uma bilbioteca específica para tratamento de dados.

In [11]:
import numpy as np
import urllib
import urllib.request

Vamos criar um objeto chamado url que terá a string do dataset.

In [12]:
with urllib.request.urlopen("http://goo.gl/j0Rvxq") as url:
    s = url.read()

O objeto 'raw_data' vai abrir uma url e vai armazenar os dados.

In [13]:
raw_data = s

Com o objeto armazenado, vamos realizar a atribuição, já realizar a separaçnao pelo delimitador que será uma vírgula.

In [14]:
dataset = np.loadtxt(raw_data, delimiter=",")

Agora vamos conferir o nosso dataset como está.

In [15]:
print(dataset.shape)

(23279,)


Aos moldes do exemplo anterior, vamos atribuir as variáveis independentes no objeto X e a variável dependente no objeto y; sendo que:

- Da variável 0 até a 7 serão as variáveis independentes (X)
- A variável 8 será a variável dependente.



In [16]:
X = dataset[:,0:7]

IndexError: too many indices for array

In [17]:
y = dataset[:,8]

IndexError: too many indices for array

Dados carregados e checados, vamos agora realizar algumas atividades de pré-processamento. 

A primeira de pré-processamento que vamos utilizar será o Scaling. 

O Scaling é bastante utilizado quando se tem um conjunto de dados com valores com uma variância muito alta, e que em termos custo computacional seria custoso.



Primeiramente, vamos importar a biblioteca 'preprocessing' do pacote do Scikit-Learn. 

In [18]:
from sklearn import preprocessing

Vamos criar um objeto chamado 'normalized_X' em que vamos aplicar a função scale da biblioteca preprocessing. 

In [19]:
normalized_X = preprocessing.scale(X)

Objeto criado, vamos agora fazer uma pequena comparação entre o objeto X (nossos dados) e o objeto normalized_X (o que acabamos de normalizar).

In [20]:
X

array([[ 5.1,  3.5,  1.4,  0.2],
       [ 4.9,  3. ,  1.4,  0.2],
       [ 4.7,  3.2,  1.3,  0.2],
       [ 4.6,  3.1,  1.5,  0.2],
       [ 5. ,  3.6,  1.4,  0.2],
       [ 5.4,  3.9,  1.7,  0.4],
       [ 4.6,  3.4,  1.4,  0.3],
       [ 5. ,  3.4,  1.5,  0.2],
       [ 4.4,  2.9,  1.4,  0.2],
       [ 4.9,  3.1,  1.5,  0.1],
       [ 5.4,  3.7,  1.5,  0.2],
       [ 4.8,  3.4,  1.6,  0.2],
       [ 4.8,  3. ,  1.4,  0.1],
       [ 4.3,  3. ,  1.1,  0.1],
       [ 5.8,  4. ,  1.2,  0.2],
       [ 5.7,  4.4,  1.5,  0.4],
       [ 5.4,  3.9,  1.3,  0.4],
       [ 5.1,  3.5,  1.4,  0.3],
       [ 5.7,  3.8,  1.7,  0.3],
       [ 5.1,  3.8,  1.5,  0.3],
       [ 5.4,  3.4,  1.7,  0.2],
       [ 5.1,  3.7,  1.5,  0.4],
       [ 4.6,  3.6,  1. ,  0.2],
       [ 5.1,  3.3,  1.7,  0.5],
       [ 4.8,  3.4,  1.9,  0.2],
       [ 5. ,  3. ,  1.6,  0.2],
       [ 5. ,  3.4,  1.6,  0.4],
       [ 5.2,  3.5,  1.5,  0.2],
       [ 5.2,  3.4,  1.4,  0.2],
       [ 4.7,  3.2,  1.6,  0.2],
       [ 4

In [21]:
normalized_X

array([[ -9.00681170e-01,   1.03205722e+00,  -1.34127240e+00,
         -1.31297673e+00],
       [ -1.14301691e+00,  -1.24957601e-01,  -1.34127240e+00,
         -1.31297673e+00],
       [ -1.38535265e+00,   3.37848329e-01,  -1.39813811e+00,
         -1.31297673e+00],
       [ -1.50652052e+00,   1.06445364e-01,  -1.28440670e+00,
         -1.31297673e+00],
       [ -1.02184904e+00,   1.26346019e+00,  -1.34127240e+00,
         -1.31297673e+00],
       [ -5.37177559e-01,   1.95766909e+00,  -1.17067529e+00,
         -1.05003079e+00],
       [ -1.50652052e+00,   8.00654259e-01,  -1.34127240e+00,
         -1.18150376e+00],
       [ -1.02184904e+00,   8.00654259e-01,  -1.28440670e+00,
         -1.31297673e+00],
       [ -1.74885626e+00,  -3.56360566e-01,  -1.34127240e+00,
         -1.31297673e+00],
       [ -1.14301691e+00,   1.06445364e-01,  -1.28440670e+00,
         -1.44444970e+00],
       [ -5.37177559e-01,   1.49486315e+00,  -1.28440670e+00,
         -1.31297673e+00],
       [ -1.26418478e

Como podemos atestar visualmente, tivemos uma mudança significativa de dos valores, em que  praticamente os valores sairam da casa da metade de dezena, para um nível muito maior de precisão decimal e também possuí valores negativos.


Outra forma de pré-processamento é a normalização.

A normalização tem o mesmo princípio que a padronização, porém com a diferença fundamental que o nível de escala é reduzido.

Da mesma forma que fizemos na padronização, vamos criar um objeto e vamos usar a função 'normalize' da biblioteca preprocessing. 

In [22]:
standarized_X = preprocessing.normalize(X)

Vamos agora comparar o output dos dados dos objetos X e standarized_X.

In [23]:
X

array([[ 5.1,  3.5,  1.4,  0.2],
       [ 4.9,  3. ,  1.4,  0.2],
       [ 4.7,  3.2,  1.3,  0.2],
       [ 4.6,  3.1,  1.5,  0.2],
       [ 5. ,  3.6,  1.4,  0.2],
       [ 5.4,  3.9,  1.7,  0.4],
       [ 4.6,  3.4,  1.4,  0.3],
       [ 5. ,  3.4,  1.5,  0.2],
       [ 4.4,  2.9,  1.4,  0.2],
       [ 4.9,  3.1,  1.5,  0.1],
       [ 5.4,  3.7,  1.5,  0.2],
       [ 4.8,  3.4,  1.6,  0.2],
       [ 4.8,  3. ,  1.4,  0.1],
       [ 4.3,  3. ,  1.1,  0.1],
       [ 5.8,  4. ,  1.2,  0.2],
       [ 5.7,  4.4,  1.5,  0.4],
       [ 5.4,  3.9,  1.3,  0.4],
       [ 5.1,  3.5,  1.4,  0.3],
       [ 5.7,  3.8,  1.7,  0.3],
       [ 5.1,  3.8,  1.5,  0.3],
       [ 5.4,  3.4,  1.7,  0.2],
       [ 5.1,  3.7,  1.5,  0.4],
       [ 4.6,  3.6,  1. ,  0.2],
       [ 5.1,  3.3,  1.7,  0.5],
       [ 4.8,  3.4,  1.9,  0.2],
       [ 5. ,  3. ,  1.6,  0.2],
       [ 5. ,  3.4,  1.6,  0.4],
       [ 5.2,  3.5,  1.5,  0.2],
       [ 5.2,  3.4,  1.4,  0.2],
       [ 4.7,  3.2,  1.6,  0.2],
       [ 4

In [24]:
standarized_X

array([[ 0.80377277,  0.55160877,  0.22064351,  0.0315205 ],
       [ 0.82813287,  0.50702013,  0.23660939,  0.03380134],
       [ 0.80533308,  0.54831188,  0.2227517 ,  0.03426949],
       [ 0.80003025,  0.53915082,  0.26087943,  0.03478392],
       [ 0.790965  ,  0.5694948 ,  0.2214702 ,  0.0316386 ],
       [ 0.78417499,  0.5663486 ,  0.2468699 ,  0.05808704],
       [ 0.78010936,  0.57660257,  0.23742459,  0.0508767 ],
       [ 0.80218492,  0.54548574,  0.24065548,  0.0320874 ],
       [ 0.80642366,  0.5315065 ,  0.25658935,  0.03665562],
       [ 0.81803119,  0.51752994,  0.25041771,  0.01669451],
       [ 0.80373519,  0.55070744,  0.22325977,  0.02976797],
       [ 0.786991  ,  0.55745196,  0.26233033,  0.03279129],
       [ 0.82307218,  0.51442011,  0.24006272,  0.01714734],
       [ 0.8025126 ,  0.55989251,  0.20529392,  0.01866308],
       [ 0.81120865,  0.55945424,  0.16783627,  0.02797271],
       [ 0.77381111,  0.59732787,  0.2036345 ,  0.05430253],
       [ 0.79428944,  0.

Todos os dados foram para a escala de 0 a 1. 

Com isso os algoritmos podem fazer um trabalho muito melhor, dado que o range de dados foi diminuído. 

Agora, um aspecto fundamental quando lidamos com um grande volume de dados em que diz respeito ao número de atributos (ou dimensionalidade) é saber quais variáveis independentes são importantes para treinamento do modelo de predição. 

Para fazer essa atividade, vamos usar uma das bibliotecas do Scikit chamada 'ExtraTreesClassifier' que tem uma função que atesta quais variáveis tem o maior poder preditivo. 

Primeiramente, vamos importar a biblioteca.

In [25]:
from sklearn.ensemble import ExtraTreesClassifier

Biblioteca importada, agora vamos criar o nosso modelo realizando uma chamada na biblioteca.

In [26]:
model = ExtraTreesClassifier()

Agora vamos realizar o 'fitting' do modelo. 

Uma dica é que sempre que for ser utilizado algum modelo para fitting, usando bibliotecas de aprendizado supervisionado, a regra sempre será:

- objeto_do_modelo.fit (X,y)

Em que objeto_do_modelo é o modelo preditivo utilizado, X as variáveis independentes e y a variável dependente a ser predita.

In [27]:
model.fit(datasets.data, datasets.target)

ExtraTreesClassifier(bootstrap=False, class_weight=None, criterion='gini',
           max_depth=None, max_features='auto', max_leaf_nodes=None,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
           oob_score=False, random_state=None, verbose=0, warm_start=False)

Ajustamos o modelo, porém vamos usar as configurações básicas. Mas de maneira geral usamos o algoritmo de árvores de decisão para obtermos a métrica de importância de cada variável.

In [28]:
print(model.feature_importances_)

[ 0.04681289  0.04088293  0.46326696  0.44903722]


In [29]:
model.fit(X, y)

ExtraTreesClassifier(bootstrap=False, class_weight=None, criterion='gini',
           max_depth=None, max_features='auto', max_leaf_nodes=None,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
           oob_score=False, random_state=None, verbose=0, warm_start=False)

In [30]:
print(model.feature_importances_)

[ 0.13100367  0.05959566  0.39799697  0.4114037 ]


Esse resultado mostra que em termos de importância preditiva, as variáveis mais significativas são a 3(57%), 4(33%), 1 e (5% cada).

Caso houvessem muitas variáveis a serem consideradas, e quisessemos diminuir o custo computacional (e consequentemente o tempo) poderiamos eliminar essas duas variáveis, mas como estamos simplificando vamos considerar tudo.

Com isso terminamos o pré-processamento básico. 

Como regra geral de modelagem, o ideal é que essas informações venham já processadas do banco de dados, dado que o mecanismo do SGBD é apto para esse tipo de atividade, tanto em termos de velocidade quanto em robustez de processamento. 

E além do mais, o importante é que quanto mais modularizado em que há uma cadeia de ferramentas fazendo aquilo o que ela faz de melhor funcionando, a complexidade do sistema como um todo caí, e aumenta a robuztez quanto à problemas que possam acontecer. 